File: Backend/Services/UsuarioService.cs
using Backend.Data;
using Backend.DTOs.Common;
using Backend.DTOs.Usuarios;
using Backend.Models;
using Backend.Services.Interfaces;
using Microsoft.EntityFrameworkCore;

namespace Backend.Services;

public class UsuarioService : IUsuarioService
{
    private readonly AppDbContext _context;
    private readonly IAuthService _authService;

    public UsuarioService(AppDbContext context, IAuthService authService)
    {
        _context = context;
        _authService = authService;
    }

    public async Task<ResultadoPaginadoDto<UsuarioDto>> ObterTodosAsync(PaginacaoDto paginacao)
    {
        var query = _context.Usuarios.AsQueryable();

        var total = await query.CountAsync();

        var usuarios = await query
            .OrderByDescending(u => u.DataCriacao)
            .Skip((paginacao.Pagina - 1) * paginacao.TamanhoPagina)
            .Take(paginacao.TamanhoPagina)
            .Select(u => MapToDto(u))
            .ToListAsync();

        return new ResultadoPaginadoDto<UsuarioDto>
        {
            Dados = usuarios,
            PaginaAtual = paginacao.Pagina,
            TamanhoPagina = paginacao.TamanhoPagina,
            TotalItens = total,
            TotalPaginas = (int)Math.Ceiling(total / (double)paginacao.TamanhoPagina)
        };
    }

    public async Task<UsuarioDto?> ObterPorIdAsync(int id)
    {
        var usuario = await _context.Usuarios.FindAsync(id);
        return usuario == null ? null : MapToDto(usuario);
    }

    public async Task<UsuarioDto> CriarAsync(CriarUsuarioDto dto)
    {
        var usuario = new Usuario
        {
            Nome = dto.Nome,
            Email = dto.Email,
            SenhaHash = _authService.HashSenha(dto.Senha),
            Perfil = dto.Perfil,
            DataCriacao = DateTime.UtcNow,
            Ativo = true
        };

        _context.Usuarios.Add(usuario);
        await _context.SaveChangesAsync();

        return MapToDto(usuario);
    }

    public async Task<UsuarioDto?> AtualizarAsync(int id, AtualizarUsuarioDto dto)
    {
        var usuario = await _context.Usuarios.FindAsync(id);
        if (usuario == null) return null;

        usuario.Nome = dto.Nome;
        usuario.Email = dto.Email;
        usuario.Perfil = dto.Perfil;
        usuario.Ativo = dto.Ativo;

        await _context.SaveChangesAsync();

        return MapToDto(usuario);
    }

    public async Task<bool> ExcluirAsync(int id, int usuarioLogadoId)
    {
        if (id == usuarioLogadoId)
            throw new InvalidOperationException("Você não pode excluir a si mesmo.");

        var usuario = await _context.Usuarios.FindAsync(id);
        if (usuario == null) return false;

        _context.Usuarios.Remove(usuario);
        await _context.SaveChangesAsync();

        return true;
    }

    public async Task<bool> AlterarSenhaAsync(int id, AlterarSenhaDto dto)
    {
        var usuario = await _context.Usuarios.FindAsync(id);
        if (usuario == null) return false;

        if (!_authService.VerificarSenha(dto.SenhaAtual, usuario.SenhaHash))
            throw new InvalidOperationException("Senha atual incorreta.");

        usuario.SenhaHash = _authService.HashSenha(dto.NovaSenha);
        await _context.SaveChangesAsync();

        return true;
    }

    private static UsuarioDto MapToDto(Usuario usuario) => new()
    {
        Id = usuario.Id,
        Nome = usuario.Nome,
        Email = usuario.Email,
        Perfil = usuario.Perfil.ToString(),
        DataCriacao = usuario.DataCriacao,
        Ativo = usuario.Ativo
    };
}

File: backend/src/services/usuario.service.ts
import { supabase } from '../config/database';
import { PerfilUsuario, PerfilUsuarioLabel } from '../models/enums';
import {
  UsuarioDto,
  CriarUsuarioDto,
  AtualizarUsuarioDto,
  AlterarSenhaDto,
} from '../dtos/usuario.dto';
import { PaginacaoDto, ResultadoPaginadoDto, criarResultadoPaginado } from '../dtos/common.dto';
import { hashSenha, verificarSenha } from './auth.service';
import { InvalidOperationError } from '../middlewares/error.middleware';

// ─── Mapper ──────────────────────────────────────────────────────────
function mapToDto(u: any): UsuarioDto {
  return {
    id: u.id,
    nome: u.nome,
    email: u.email,
    perfil: PerfilUsuarioLabel[u.perfil as PerfilUsuario] || u.perfil.toString(),
    dataCriacao: u.data_criacao,
    ativo: u.ativo,
  };
}

// ─── Listar (paginado) ──────────────────────────────────────────────
export async function obterTodosAsync(
  paginacao: PaginacaoDto,
): Promise<ResultadoPaginadoDto<UsuarioDto>> {
  const { data, count, error } = await supabase
    .from('usuarios')
    .select('id, nome, email, perfil, data_criacao, ativo', { count: 'exact' })
    .order('data_criacao', { ascending: false })
    .range(
      (paginacao.pagina - 1) * paginacao.tamanhoPagina,
      paginacao.pagina * paginacao.tamanhoPagina - 1,
    );

  if (error) throw new Error(error.message);

  const dados = (data || []).map(mapToDto);
  return criarResultadoPaginado(dados, paginacao.pagina, paginacao.tamanhoPagina, count || 0);
}

// ─── Buscar por ID ───────────────────────────────────────────────────
export async function obterPorIdAsync(id: number): Promise<UsuarioDto | null> {
  const { data, error } = await supabase
    .from('usuarios')
    .select('id, nome, email, perfil, data_criacao, ativo')
    .eq('id', id)
    .single();

  if (error || !data) return null;
  return mapToDto(data);
}

// ─── Criar ───────────────────────────────────────────────────────────
export async function criarAsync(dto: CriarUsuarioDto): Promise<UsuarioDto> {
  // Validação de email único
  const { data: existente } = await supabase
    .from('usuarios')
    .select('id')
    .eq('email', dto.email)
    .maybeSingle();

  if (existente) {
    throw new InvalidOperationError('Já existe um usuário cadastrado com este email.');
  }

  const senhaHasheada = await hashSenha(dto.senha);

  const { data, error } = await supabase
    .from('usuarios')
    .insert({
      nome: dto.nome,
      email: dto.email,
      senha_hash: senhaHasheada,
      perfil: dto.perfil,
      ativo: true,
    })
    .select('id, nome, email, perfil, data_criacao, ativo')
    .single();

  if (error) throw new Error(error.message);
  return mapToDto(data);
}

// ─── Atualizar ───────────────────────────────────────────────────────
export async function atualizarAsync(
  id: number,
  dto: AtualizarUsuarioDto,
): Promise<UsuarioDto | null> {
  const { data, error } = await supabase
    .from('usuarios')
    .update({
      nome: dto.nome,
      email: dto.email,
      perfil: dto.perfil,
      ativo: dto.ativo,
    })
    .eq('id', id)
    .select('id, nome, email, perfil, data_criacao, ativo')
    .single();

  if (error) return null;
  return mapToDto(data);
}

// ─── Excluir (com proteção contra auto-exclusão) ─────────────────────
export async function excluirAsync(id: number, usuarioLogadoId: number): Promise<boolean> {
  if (id === usuarioLogadoId) {
    throw new InvalidOperationError('Você não pode excluir a si mesmo.');
  }

  const { data, error } = await supabase
    .from('usuarios')
    .delete()
    .eq('id', id)
    .select('id')
    .single();

  if (error || !data) return false;
  return true;
}

// ─── Alterar Senha ───────────────────────────────────────────────────
export async function alterarSenhaAsync(id: number, dto: AlterarSenhaDto): Promise<boolean> {
  const { data: usuario, error } = await supabase
    .from('usuarios')
    .select('id, senha_hash')
    .eq('id', id)
    .single();

  if (error || !usuario) return false;

  const senhaCorreta = await verificarSenha(dto.senhaAtual, usuario.senha_hash);
  if (!senhaCorreta) {
    throw new InvalidOperationError('Senha atual incorreta.');
  }

  const novaSenhaHash = await hashSenha(dto.novaSenha);

  const { error: updateError } = await supabase
    .from('usuarios')
    .update({ senha_hash: novaSenhaHash })
    .eq('id', id);

  if (updateError) throw new Error(updateError.message);
  return true;
}
